typing/rec_check.ml

```diff
@@ -306,338 +322,671 @@ let classify_expression : Typedtree.expression -> sd =
+  fun f li m ->
+    List.fold_left (fun env item -> Env.join env (f item m)) Env.empty li
+let array : 'a. ('a -> term_judg) -> 'a array -> term_judg =
+  fun f ar m ->
+    Array.fold_left (fun env item -> Env.join env (f item m)) Env.empty ar
+
+let single : Ident.t -> term_judg = Env.single
+let remove_id : Ident.t -> term_judg -> term_judg =
+  fun id f m -> Env.remove id (f m)
+let remove_ids : Ident.t list -> term_judg -> term_judg =
+  fun ids f m -> Env.remove_list ids (f m)
+
+let join : term_judg list -> term_judg =
+  fun li m -> Env.join_list (List.map (fun f -> f m) li)
+
+let empty = fun _ -> Env.empty
+
+(* A judgment [judg] takes a mode from the context as input, and
+   returns an environment. The judgment [judg << m], given a mode [m']
+   from the context, evaluates [judg] in the composed mode [m'[m]]. *)
+let (<<) : term_judg -> Mode.t -> term_judg =
+  fun f inner_mode -> fun outer_mode -> f (Mode.compos outer_mode inner_mode)
+
+(* A binding judgment [binder] expects a mode and an inner environment,
+   and returns an outer environment. [binder >> judg] computes
+   the inner environment as the environment returned by [judg]
+   in the ambient mode. *)
+let (>>) : bind_judg -> term_judg -> term_judg =
+  fun binder term mode -> binder mode (term mode)
+
+(* Expression judgment:
+     G |- e : m
+   where (m) is an output of the code and (G) is an output;
+   in the Prolog mode notation, this is (+G |- -e : -m).
+*)
+let rec expression : Typedtree.expression -> term_judg =
+  fun exp -> match exp.exp_desc with
     | Texp_ident (pth, _, _) ->
-        (path env pth)
+      path pth
     | Texp_let (rec_flag, bindings, body) ->
-      let env', ty = value_bindings rec_flag env bindings in
-      (* Here and in other binding constructs 'discard' is used in a
-          similar way to the way it's used in sequence: uses are
-          propagated, but unguarded access are not. *)
-      Use.join (Use.discard ty) (expression (Env.join env env') body)
-    | Texp_letmodule (x, _, m, e) ->
-      let ty = modexp env m in
-      Use.join (Use.discard ty) (expression (Ident.add x ty env) e)
+      value_bindings rec_flag bindings >> expression body
+    | Texp_letmodule (x, _, mexp, e) ->
+      module_binding (x, mexp) >> expression e
     | Texp_match (e, cases, _) ->
-      let t = expression env e in
-      list (case ~scrutinee:t) env cases
-    | Texp_for (_, _, e1, e2, _, e3) ->
-      Use.(join
-              (join
-                  (inspect (expression env e1))
-                  (inspect (expression env e2)))
-              (* The body is evaluated, but not used, and not available
-                  for inclusion in another value *)
-              (discard (expression env e3)))
-
+      (*
+         (Gi; mi |- pi -> ei : m)^i
+         G |- e : sum(mi)^i
+         ----------------------------------------------
+         G + sum(Gi)^i |- match e with (pi -> ei)^i : m
```

Comment:
A general comment on the ASCII typing rules: I'm glad to have them, but I find them quite hard to read, perhaps because there's so much syntax.

I don't expect everyone to share this view, but I'd find it much easier with a bit of Unicode:

         (Γᵢ; mᵢ ⊢ pᵢ → eᵢ : m)ⁱ
         Γ ⊢ e : sum(mᵢ)ⁱ
         ──────────────────────────────────────────
         Γ + sum(Γᵢ)ⁱ ⊢ match e with (pᵢ → eᵢ)ⁱ : m

Reply:
I could do Unicode but my understanding was that we are not currently supposed to use them (especially outside string literals), for portability/whatever reasons (cc @damiendoligez ?). I think it's better to stay ASCII-only for now.